Add mode to make_temp

dubiousjim opened the pull request:
This is part of a series of PRs towards #4017.

This PR adds `mode` option for makeTempDir and makeTempFile. As with mkdir, this
argument is only used when a new dir or file is created. Defaults to
0o700 for directories and 0o600 for files.

We make the tests verify that the `mode` used is subject to the
process's umask (here just assumed to be the default 0o022 on Unix, or 0
on Windows).

ry commented on cli/js/lib.deno.ns.d.ts:
```diff
@@ -778,6 +778,10 @@ declare namespace Deno {
     suffix?: string;
+    /** Permissions to use when creating the file or directory (defaults to
+     * `0o700`, before the process's umask).
+     * Does nothing/raises on Windows. */
```
Does it do nothing or does it raise on Windows? Could you clarify the comment?

dubiousjim replied:
I hoped to prompt you (or others on the dev team) to weigh in on what you wanted to do in such cases. Noop or throw? Somewhere I have a list of all the functionality I came across where this choice needed to be made; I'll look for it and make an issue. It'd be good to have an across-the-board decision. My inclination would be to throw.

ry replied:
I think in this case no-op would be best bet...

ry commented on cli/js/make_temp_test.ts:
```diff
@@ -26,6 +26,19 @@ unitTest({ perms: { write: true } }, function makeTempDirSyncSuccess(): void {
 });
 
+unitTest(
+  { perms: { read: true, write: true } },
```
Use 
```ts
{
  perms: { read: true, write: true },
  skip: Deno.build.os == "win",
},
```

ry commented on cli/js/make_temp_test.ts:
```diff
@@ -26,6 +26,19 @@ unitTest({ perms: { write: true } }, function makeTempDirSyncSuccess(): void {
+      // Skip windows
+      // assertEquals(pathInfo.mode, 0o737 & ~Deno.umask());
+      assertEquals(pathInfo.mode & 0o777, 0o715); // assume umask 0o022
+    }
```
You should use `if (Deno.build.os == "win")` and test accordingly. The way this is now, it could be that Linux returns null for mode, and the test would pass. Also please don't leave commented out code.

dubiousjim replied:
Ok, I was following the lead of the existing tests for mkdir. Happy to make all these checks more explicit.

Happy also to delete the commented code, but which test behavior do you prefer: just assume the umask is 0o022 (as is true on the test machines)? Or explicitly query the umask? (Then these are less "unit" tests.)

ry replied:
I think explicitly querying sounds better to me.

dubiousjim commented:
This isn't ready to be merged... [two issues](https://github.com/denoland/deno/pull/4286#discussion_r390191212) need to be resolved. On noop vs throw (see #4306), @ry prefers noop but @kitsonk prefers throwing.

dubiousjim commented:
This could be ready to go: I addressed @ry's two comments. But I implemented it by throwing on Windows if an explicit `mode` was provided. For two reasons: 1. it's easier to reverse that and switch to noop than the other way, and 2. the recent umask(mode) PR went with throwing on Windows, and as I said in #4306, I hoped for a consistent policy, to guide how to proceed when implementing new functionality. But it'd be easy, and I'm happy, to change this to make WIndows silently ignore an explicit `mode` if that's what's wanted.

ry commented on cli/js/lib.deno.ns.d.ts:
```diff
@@ -786,6 +786,10 @@ declare namespace Deno {
     suffix?: string;
+    /** Permissions to use when creating the file or directory (defaults to
+     * `0o700`, before the process's umask).
+     * Raises on Windows. */
```
It doesn't raise on windows.

ry commented on cli/js/tests/make_temp_test.ts:
```diff
@@ -26,6 +26,15 @@ unitTest({ perms: { write: true } }, function makeTempDirSyncSuccess(): void {
 });
 
+unitTest(
+  { skip: Deno.build.os === "win", perms: { read: true, write: true } },
```
Don't skip windows.

ry commented on cli/js/tests/make_temp_test.ts:
```diff
@@ -26,6 +26,15 @@ unitTest({ perms: { write: true } }, function makeTempDirSyncSuccess(): void {
+  function makeTempDirSyncMode(): void {
+    const path = Deno.makeTempDirSync({ mode: 0o737 });
+    const pathInfo = Deno.statSync(path);
+    assertEquals(pathInfo.mode! & 0o777, 0o737 & ~Deno.umask());
```
Skip only this assert on windows.

ry commented on cli/fs.rs:
```diff
@@ -76,15 +77,17 @@ pub fn make_temp(
+      let mut open_options = OpenOptions::new();
+      open_options.write(true).create_new(true);
+      #[cfg(unix)]
+      open_options.mode(mode & 0o777);
```
Good

ry commented:
Now I'm looking through the Go code, they don't have an option for specifying permissions for temp dir. 
https://golang.org/pkg/io/ioutil/#TempDir

So I don't think this patch is useful...

ry commented:
@dubiousjim Is this a feature you need or are you adding this preemptively?

dubiousjim commented:
Fair enough, I just checked and Node doesn't have the option either. I don't have a use case. (I didn't set out thinking "we need this feature, let's create a PR." The PR got distilled out of some broad hacking on the fs ops that happened to add mode here when it added it elsewhere.)

I'd be happy with us simply using 0o700 for makeTempDir and 0o600 for makeTempFile (what this PR makes the defaults). Less restrictive perms would be bad.

Eventually, I'd prefer a version of makeTempFile/Dir that return rids to opened files/dirs anyway, then if one needed different perms one could just use fchmod (which is a bit further along in the queue of PRs I'm pushing) on that. But handling rids to opened dirs has to wait for more infrastructure to be in place.

If you want to close this PR, I don't mind. But there's still some useful stuff in here even if we eliminate the `mode` option (for example, the 0o700/0o600 defaults) and it'd be easier on my side for me to amend this branch to remove the `mode` option, and edit the PR's description. Of course it's your call.

ry commented:
> Eventually, I'd prefer a version of makeTempFile/Dir that return rids to opened files/dirs

Yes that sounds useful.

> But there's still some useful stuff in here even if we eliminate the mode option (for example, the 0o700/0o600 defaults

I think it would be better to start a new PR for that. You can just squash your branch and remove the make_temp dir changes.


Ok I'm closing this PR. I appreciate the effort as always! The code base improves by having more eyes on it even if changes don't ultimately land.
